### PR TITLE
Feature/integrate live data

### PR DESCRIPTION
**Pull request by CBarrett2**
Integrates live data option into chart

### CODE REVIEW

**Comment by CBarrett2** [comment]
Resolves #89 and resolves #59

File: src/assets/js/main.js [hunk 1]
```diff
@@ -11,6 +11,16 @@
         return visibleData;
     }
 
+    function padExtent(extent, period) {
```
Comment (jleft) [inline src/assets/js/main.js]: Feels more like a util to me? Consider factoring this out into its own component?

Comment (CBarrett2) [inline src/assets/js/main.js]: Like in a seperate file or as a seperate PR? I implemented this here so a single data point would display properly in the live data demo

Comment (jleft) [inline src/assets/js/main.js]: > Like in a seperate file or as a seperate PR?

Both, as I believe this relates to more than just a single data point. e.g. the first and last data points are cut off in the current implementation #89

Comment (CBarrett2) [inline src/assets/js/main.js]: Cool, I'll do that

File: src/assets/js/main.js [hunk 2]
```diff
@@ -114,23 +102,133 @@
...
+        d3.select('#period-span').style('visibility', visibility);
+        d3.select('#product-span').style('visibility', visibility);
+    }
+
+    d3.select('#type-selection')
+        .on('change', function() {
+            var type = d3.select(this).property('value');
+            if (type === 'live') {
+                toggleLiveFeedUI(true);
+                currData = [];
```
Comment (tlsim) [inline src/assets/js/main.js]: These 4 lines are repeated elsewhere - maybe extract into a function?

File: src/assets/js/main.js [hunk 3]
```diff
@@ -337,14 +402,68 @@
         selection.call(navTimeSeries);
     };
 
+    function hideOrShowChartsAndSetLoadingText(hidden, text) {
```
Comment (tlsim) [inline src/assets/js/main.js]: No need for the inner function - just call setLoadingText and toggleVisibility separately.

File: src/assets/js/main.js [hunk 4]
```diff
@@ -114,23 +102,133 @@
...
 
     var gridlines = fc.annotation.gridline()
         .yTicks(5)
         .xTicks(0);
 
+    // Create and apply the Moving Average
+    var movingAverage = fc.indicator.algorithm.movingAverage();
+
+    // Create a line that renders the result
+    var ma = fc.series.line()
```
Comment (tlsim) [inline src/assets/js/main.js]: movingAverageLine

File: src/index.html [hunk 5]
```diff
@@ -48,8 +49,37 @@
...
+            <option value="fake">Data generator</option>
+            <option value="live">Live data</option>
+          </select>
+        </div>
+        <div class="col-md-2">
+          <span id="period-span">
+            <select id="period-selection" class="form-control">
+              <option value="43200">12hr</option>
+              <option value="3600">1hr</option>
+              <option value="50">1min</option>
```
Comment (tlsim) [inline src/index.html]: 60 seconds?

File: src/assets/js/main.js [hunk 6]
```diff
@@ -114,23 +102,133 @@
...
         });
 
-    // Set Reset button event
+    var ohlcConverter = sc.data.feed.coinbase.ohlcWebSocketAdaptor();
+    // Default to generated data
+    var currData = fc.data.random.financial()(250);
+
+    function liveCallback(event, latestBasket) {
+        if (!event && latestBasket) {
+            if (!currData.length) {
```
Comment (tlsim) [inline src/assets/js/main.js]: Extract function for 'new basket received'.

File: src/index.html [hunk 7]
```diff
@@ -48,8 +49,37 @@
...
+          
+          <select id="type-selection" selected="type" class="form-control">
+            <option value="fake">Data generator</option>
+            <option value="live">Live data</option>
+          </select>
+        </div>
+        <div class="col-md-2">
+          <span id="period-span">
+            <select id="period-selection" class="form-control">
+              <option value="43200">12hr</option>
```
Comment (tlsim) [inline src/index.html]: Might be worth extracting period selection into a component (future task).

**Comment by CBarrett2** [comment]
Thanks for the feedback! :)

File: src/assets/js/main.js [hunk 8]
```diff
@@ -239,13 +285,27 @@
...
+                    return data;
+            }
+        });
+
         // Scale y axis
         var yExtent = fc.util.extent(getVisibleData(data, timeSeries.xDomain()), ['low', 'high']);
         // Add 10% either side of extreme high/lows
         var variance = yExtent[1] - yExtent[0];
+        if (variance === 0) {
+            // Bit of a hack to get single-value candlesticks to show nicely.
```
Comment (tlsim) [inline src/assets/js/main.js]: Looks like padExtent but for a non-time domain.

**Comment by tlsim** [comment]
Good job - the 1 min live updated candles look nice!
All comments are fairly minor code style things.
Worth thinking about the data part of #70 after this I think.

File: src/assets/js/main.js (hunk 8 above)
Comment (CBarrett2) [inline src/assets/js/main.js]: That's a great idea

**Pull request closed without merging**